### ISSUE TITLE
Multiple instances of CustomTwilioVideoView

### ISSUE THREAD

**Issue by timmeh-m** [blackuy/react-native-twilio-video-webrtc#95]
So I'm using react-navigation with twilio video. When a user presses a button, it navigates to the video call screen and a room is created. The app can go to the background and in the console there is one instance of 'onHostPause'. After a while, they eventually end the call, disconnect() is called, and they are navigated back to the previous screen. Now the issue is if they press the button again to create a room, an instance of CustomTwilioVideoView is created but the previous instance appears to still exist. When the app goes to the background, this time there are two instances of 'onHostPause'. I'm wondering if there's any way to dealloc any previous instances of CustomTwilioVideoView so as to prevent multiple copies of the same class. I have tried creating a singleton of it and returning the singleton in CustomTwiloVideoViewManager, but it doesn't completely get me there. I'd like a way to completely remove CustomTwilioVideoView when disconnect() is called until a new call is made so that onCreate is called each time. Has anyone else had similar thoughts on this?

**Comment by slycoder** [comment]
In principle, the View should disappear whenever it is unmounted from the react hierarchy.  Can you verify that the component is being unmounted (for example, monitoring componentWillUnmount on its container).  I've definitely had it happen to me when not being careful with react navigation that instead of popping a screen off the stack I'd end up creating a whole new one on top of the old one.

**Comment by Andresagustin** [comment]
@slycoder it seems that react-navigation may be the issue. I read somewhere that we need to bind the method in the constructor. How can we do this? 
The issue happens when we terminate the current cal and navigate away from the call screen, then if we go back to the call screen and try to make another call we got undefined is not an object.... _this.refs.twillioVideo

**Comment by slycoder** [comment]
I don't quite understand what you're saying.  React native and react native will totally unmount screens when navigating away.  HOWEVER, if you're using a stack navigator any screens still in the stack will never be unmounted.  If you are always pushing more screens onto the stack, rather than popping / resetting appropriately you're going to have a bad time.

**Comment by Andresagustin** [comment]
Thanks @slycoder for your reply. I'm now resetting the screen to the main one per your suggestion. But when i open again the call screen (I'm seeing the mount/unmount correctly) but the second time I got: 

![img_6264](https://user-images.githubusercontent.com/40077694/43023357-f3d2198c-8c40-11e8-84d7-2f451a8aa4ea.PNG)

this happens when I  call     this.refs.twilioVideo.connect({ roomName: this.state.channelID, accessToken: this.state.channelToken })

**Comment by Andresagustin** [comment]
sure, thanks for your help :) @slycoder :)

`Sure, here is the code i'm using

`
import React, { Component } from 'react';
const io = require('socket.io-client');

const socket = io('http://192.168.0.114:4000', {
    transports: ['websocket'],
  });
  console.disableYellowBox = true
import {
  AppRegistry,
  StyleSheet,
  Text,
  TextInput,
  View,
  Button,
  TouchableOpacity,
  Dimensions,
  TouchableHighlight
} from 'react-native';

import {
  TwilioVideoLocalView,
  TwilioVideoParticipantView,
  TwilioVideo
} from 'react-native-twilio-video-webrtc'

import EndConsultation from "./../endconsultation";

import { StackActions, NavigationActions } from 'react-navigation';

const resetAction = StackActions.pop({
  n: 1,
});

var Environment = require('../../components/config.js')
var keyURL = Environment.KEY_URL

const styles = StyleSheet.create({
  ... all my styles ...
});

export default class Example extends Component {

  componentWillUnmount(){
    alert("call screen unmount")
  }
componentWillMount(){
  alert("call screen mounted")
  var channelID = this.props.navigation.state.params.id;
  var channelToken = this.props.navigation.state.params.token;

  this.setState({status: 'connecting',
  channelID: this.props.navigation.state.params.id,
  channelToken: this.props.navigation.state.params.token,

})

        this._onConnecttoVideo()
}

_onConnecttoVideo = () =>{

    this.refs.twilioVideo.connect({ roomName: this.state.channelID, accessToken: this.state.channelToken })
    this.setState({status: 'connecting'})
  }



  state = {
    isAudioEnabled: true,
    isVideoEnabled: true,
    status: 'disconnected',
    participants: new Map(),
    videoTracks: new Map(),
  }

  _onEndButtonPress = () => {
    this.refs.twilioVideo.disconnect()
      this.props.navigation.dispatch(resetAction);
  }

  _onMuteButtonPress = () => {
    this.refs.twilioVideo.setLocalAudioEnabled(!this.state.isAudioEnabled)
      .then(isEnabled => this.setState({isAudioEnabled: isEnabled}))
  }

  _onFlipButtonPress = () => {
    this.refs.twilioVideo.flipCamera()
  }

  _onRoomDidConnect = () => {

  }

  _onRoomDidDisconnect = ({roomName, error}) => {
    console.log("disconnected from : " + roomName + error)

    this.setState({status: 'disconnected'})
  }

  _onRoomDidFailToConnect = (error) => {
    console.log("ERROR _onRoomDidFailToConnect: ", error)

    this.setState({status: 'disconnected'})
  }

  _onRoomParticipantDidDisconnect = ({roomName, participant}) => {
    console.log("other disconnected" + roomName + participant)

  }

  _onParticipantAddedVideoTrack = ({participant, track}) => {
    console.log("onParticipantAddedVideoTrack: ", participant, track)
    this.setState({status: 'connected'})
    this.setState({
      videoTracks: new Map([
        ...this.state.videoTracks,
        [track.trackId, { ...participant, ...track }]
      ]),
    });
  }

  _onParticipantRemovedVideoTrack = ({participant, track}) => {
    console.log("onParticipantRemovedVideoTrack: ", participant, track)

    const videoTracks = this.state.videoTracks
    videoTracks.delete(track.trackId)
        this.setState({status: 'disconnected'})

    this.setState({ videoTracks: new Map([ ...videoTracks ]) });
  }

  render() {
    return (
      <View>
..... bunch of stuff
      </View>
    );
  }
}

AppRegistry.registerComponent('Example', () => Example);
``

when I navigate away after a call (disconnected). And Open the call screen again to connect to a new call I got that error. (btw, removed a bunch of code that use for websocket state sharing).

Thanks!!

**Comment by slycoder** [comment]
Wait, are you calling connect in componentWillMount?  That happens before the components are mounted so none of the refs will have been set yet.  You should be doing that in componentDidMount.

**Comment by Andresagustin** [comment]
I moved the connect to componentDidMount and still have the same output. Second call fails with the same error. any ideas @slycoder ?

**Comment by Andresagustin** [comment]
Hey @timmeh-m have you been able to fix the multiple instances issue? We tried for several days with no luck.

**Comment by Andresagustin** [comment]
Hi @slycoder do you have any working example without using this.refs? (this.refs.twilioVideo.connect). It seems that this.refs is not compatible with react navigation.

Thanks!

**Comment by slycoder** [comment]
I'm not sure why refs would not be compatible with react navigation ... they are orthogonal pieces of the stack.  Refs are a fundamental feature of react --- most of the time when things don't behave as expected it's due to a race condition or a misunderstanding of the react lifecycle and/or which components are mounted.

If you want to put your code somewhere for someone to look at that would be great.  If you want to debug yourself, you could add some logging for the lifecycle methods to ensure that things are being mounted and unmounted when you expect them to be.

**Comment by Andresagustin** [comment]
thanks @slycoder for your help. Really appreciated. I have already debugged the lifecycle and I can confirm that the component is unmounting as expected. But the issue resides when you navigate away from the current screen (props.navigate) and I try to go back again. it trows 'undefined' on _this.refs.twillioVideo.connect.

**Comment by slycoder** [comment]
And you're sure the twiliovideo component is mounted before trying to connect?

**Comment by Andresagustin** [comment]
Yes, also I have all the connect logic in componentDidMount, so the component is mounted by then.

**Comment by slycoder** [comment]
Like I said, it's difficult for me to help debug at this point without being able to see the code.  I at least don't see anything obviously wrong in the code from 18 days ago, but I also don't see how react-navigation is being used there.

**Comment by harrydema** [comment]
Any update?

**Comment by ShankyRana** [comment]
Similar issue, not working on ios but working fine on Android. Any solution for this?

**Comment by slycoder** [comment]
@ShankyRana Do you have a simple repro that you can share?
